File: lib/jvozba.js
/* global get_CV_info */
/* eslint-disable camelcase  */
/*
jvozba(["lujvo","zbasu"]) ==> [{"lujvo":"jvozba","score":5858},{"lujvo":"luvzba","score":5878},{"lujvo":"jvozbas","score":6888},{"lujvo":"luvzbas","score":6908},{"lujvo":"jvozbasu","score":7897},{"lujvo":"luvzbasu","score":7917},{"lujvo":"lujvyzba","score":8008},{"lujvo":"lujvyzbas","score":9038},{"lujvo":"lujvyzbasu","score":10047}]
*/
function jvozba (arr, forbid_la_lai_doi) {
  var candid_arr = []

  for (var i = 0; i < arr.length; i++) {
    candid_arr.push(get_candid(arr[i], /* isLast: */ i === arr.length - 1))
  }

  var answers = create_every_possibility(candid_arr).map(function (rafsi_list) {
    var result = normalize(rafsi_list)
    return {lujvo: result.join(''), score: get_lujvo_score(result)}
  }).filter(function (d) {
    var l = d.lujvo
    return !(is_cmevla(l) && forbid_la_lai_doi &&
    (l.match(/^(lai|doi)/) ||
    l.match(/[aeiouy](lai|doi)/) ||
    l.match(/^la[^u]/) || // the fact that CLL explicitly forbids two sequences `la` and `lai` signifies that `lau` is not forbidden
    l.match(/[aeiouy]la[^u]/)
    )
    )
  }).sort(function (a, b) {
    return a.score - b.score
  })

  return answers
}

function is_cmevla (valsi) {
  return valsi.length >= 1 && "aeiouy'".indexOf(valsi.charAt(valsi.length - 1)) === -1
}

function normalize (rafsi_list) {
  if (rafsi_list.length === 1) {
    throw new Error('You need at least two valsi to make a lujvo')
  }

  var input = rafsi_list.concat([]) // copy
  var result = [input.pop()] // add the final rafsi

  while (input.length)	{
    var rafsi = input.pop()
    var end = rafsi.charAt(rafsi.length - 1)
    var init = result[0].charAt(0)

    if (is_4letter(rafsi)) {
      result.unshift('y')
    } else if (get_CV_info(end) === 'C' && get_CV_info(init) === 'C' && is_permissible(end, init) === 0) {
      result.unshift('y')
    } else if (end === 'n' && ['ts', 'tc', 'dz', 'dj'].indexOf(result[0].slice(0, 2)) !== -1) {
      result.unshift('y')
    } else if (input.length === 0 && is_CVV(rafsi)) { // adapting first rafsi, which is CVV; gotta think about r-hyphen
      var hyphen = 'r'
      if (result[0].startsWith('r')) {
        hyphen = 'n'
      }

      if (rafsi_list.length > 2 || !is_CCV(result[0])) {
        result.unshift(hyphen)
      }
    } else if (input.length === 0 && is_CVC(rafsi) && is_tosmabru(rafsi, result)) {
      result.unshift('y')
    }

    result.unshift(rafsi)
  }

  return result
}

function is_tosmabru (rafsi, rest) {
  // skip if cmevla
  if (is_cmevla(rest[rest.length - 1])) { // ends with a consonant
    return
  }

  var index
  for (var i = 0; i < rest.length; i++) {
    if (is_CVC(rest[i])) continue

    index = i
    if (rest[i] === 'y' ||
      (get_CV_info(rest[i]) === 'CVCCV' && is_permissible(rest[i].charAt(2), rest[i].charAt(3)) === 2)
    ) {
      break
    // further testing
    } else {
      return false
    }
  }

  // further testing

  var tmp1 = rafsi
  var tmp2 = rest[0]
  var j = 0
  do {
    if (tmp2 === 'y') return true

    if (is_permissible(tmp1.charAt(tmp1.length - 1), tmp2.charAt(0)) !== 2) {
      return false
    }
    tmp1 = tmp2
    tmp2 = rest[++j]
  } while (j <= index)
  return true
}

function is_CVV (rafsi) {
  return (get_CV_info(rafsi) === 'CVV' ||
    get_CV_info(rafsi) === "CV'V")
}

function is_CCV (rafsi) {
  return get_CV_info(rafsi) === 'CCV'
}

function is_CVC (rafsi) {
  return get_CV_info(rafsi) === 'CVC'
}

function is_4letter (rafsi) {
  return (get_CV_info(rafsi) === 'CVCC' ||
    get_CV_info(rafsi) === 'CCVC')
}

function is_permissible (c1, c2) { // 2: initial ok; 1: ok; 0: none ok
  return {
    r: {r: 0, l: 1, n: 1, m: 1, b: 1, v: 1, d: 1, g: 1, j: 1, z: 1, s: 1, c: 1, x: 1, k: 1, t: 1, f: 1, p: 1},
    l: {r: 1, l: 0, n: 1, m: 1, b: 1, v: 1, d: 1, g: 1, j: 1, z: 1, s: 1, c: 1, x: 1, k: 1, t: 1, f: 1, p: 1},
    n: {r: 1, l: 1, n: 0, m: 1, b: 1, v: 1, d: 1, g: 1, j: 1, z: 1, s: 1, c: 1, x: 1, k: 1, t: 1, f: 1, p: 1},
    m: {r: 2, l: 2, n: 1, m: 0, b: 1, v: 1, d: 1, g: 1, j: 1, z: 0, s: 1, c: 1, x: 1, k: 1, t: 1, f: 1, p: 1},
    b: {r: 2, l: 2, n: 1, m: 1, b: 0, v: 1, d: 1, g: 1, j: 1, z: 1, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    v: {r: 2, l: 2, n: 1, m: 1, b: 1, v: 0, d: 1, g: 1, j: 1, z: 1, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    d: {r: 2, l: 1, n: 1, m: 1, b: 1, v: 1, d: 0, g: 1, j: 2, z: 2, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    g: {r: 2, l: 2, n: 1, m: 1, b: 1, v: 1, d: 1, g: 0, j: 1, z: 1, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    j: {r: 1, l: 1, n: 1, m: 2, b: 2, v: 2, d: 2, g: 2, j: 0, z: 0, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    z: {r: 1, l: 1, n: 1, m: 2, b: 2, v: 2, d: 2, g: 2, j: 0, z: 0, s: 0, c: 0, x: 0, k: 0, t: 0, f: 0, p: 0},
    s: {r: 2, l: 2, n: 2, m: 2, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 0, c: 0, x: 1, k: 2, t: 2, f: 2, p: 2},
    c: {r: 2, l: 2, n: 2, m: 2, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 0, c: 0, x: 0, k: 2, t: 2, f: 2, p: 2},
    x: {r: 2, l: 2, n: 1, m: 1, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 1, c: 0, x: 0, k: 0, t: 1, f: 1, p: 1},
    k: {r: 2, l: 2, n: 1, m: 1, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 1, c: 1, x: 0, k: 0, t: 1, f: 1, p: 1},
    t: {r: 2, l: 1, n: 1, m: 1, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 2, c: 2, x: 1, k: 1, t: 0, f: 1, p: 1},
    f: {r: 2, l: 2, n: 1, m: 1, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 1, c: 1, x: 1, k: 1, t: 1, f: 0, p: 1},
    p: {r: 2, l: 2, n: 1, m: 1, b: 0, v: 0, d: 0, g: 0, j: 0, z: 0, s: 1, c: 1, x: 1, k: 1, t: 1, f: 1, p: 0}
  }[c1][c2]
}
/*
  create_every_possibility([[1,11], [2], [3,33,333]]) ==> [ [1,2,3],[11,2,3],  [1,2,33],[11,2,33],  [1,2,333],[11,2,333] ]
  create_every_possibility([[1,11]]) ==> [ [1],[11] ]
*/
function create_every_possibility (aa) {
  var arr_arr = JSON.parse(JSON.stringify(aa))
  if (arr_arr.length === 0) {
    return [[]]
  }
  var arr = arr_arr.pop()

  var result = []
  for (var i = 0; i < arr.length; i++) {
    var e = arr[i]

    result = result.concat(create_every_possibility(arr_arr).map(function (f) {
      return f.concat([e])
    }))
  }
  return result
}

// get_candid("bloti", false) ==> ["lot", "blo", "lo'i", "blot"]
// get_candid("gismu", true) ==> ["gim", "gi'u", "gismu", "gism"]
function get_candid (selrafsi, isLast) {
  if (cmavo_rafsi_list[selrafsi]) {
    return cmavo_rafsi_list[selrafsi]
  } else if (gismu_rafsi_list[selrafsi]) {
    var gismu = selrafsi
    var candid = gismu_rafsi_list[gismu].concat([])

    if (isLast) {
      candid.push(gismu)
    }

    var chopped = gismu.slice(0, -1)
    if (chopped !== 'brod') { candid.push(chopped) }

    return candid
  } else {
    throw new Error('no rafsi for word ' + selrafsi)
  }
}

function search_selrafsi_from_rafsi2 (rafsi) {
  if (gismu_rafsi_list[rafsi]) return rafsi // 5-letter rafsi

  /*
    I spent 45 minutes trying to find out whether "brod" can be a rafsi for "brodV", but couldn't find that out.
    Thus, for the present I forbid the use of "brod" as a rafsi.
  */
  if (rafsi !== 'brod' && rafsi.length === 4 && rafsi.indexOf("'") === -1) { // 4-letter rafsi
    for (var u = 0; u < 5; u++) {
      var gismu_candid = rafsi + 'aeiou'.charAt(u)
      if (gismu_rafsi_list[gismu_candid]) return gismu_candid
    }
  }
  for (var i in gismu_rafsi_list) {
    if (gismu_rafsi_list[i].indexOf(rafsi) !== -1) return i
  }
  for (var j in cmavo_rafsi_list) {
    if (cmavo_rafsi_list[j].indexOf(rafsi) !== -1) return j
  }

  return null
}

function search_selrafsi_from_rafsi (rafsi) {
  var selrafsi = search_selrafsi_from_rafsi2(rafsi)
  if (selrafsi != null) {
    return selrafsi
  } else {
    throw new Error('no word for rafsi ' + rafsi)
  }
}

File: scripts/jvozba.py
# -*- coding: utf-8 -*-
'''lujvo 合成 tools'''
import json
import re
from util import lujvo_score, get_cv, permissible_c, is_cmevla

def jvozba(parts: list, forbid_la_lai_doi=False):
    '''
    lujvo 合成
    '''
    rafsis_ary = [get_candid(raf, i == len(parts) - 1) for i, raf in enumerate(parts)]

    def maps(rafsi_list):
        '''
        answer results
        '''
        result = normalize(rafsi_list)
        return {'lujvo': ''.join(result), 'score': lujvo_score(result)}

    answers = [maps(x) for x in create_every_possibility(rafsis_ary)]

    def la_lai_doi(lujvo):
        '''
        filter function:
        '''
        return not (
            is_cmevla(lujvo) and forbid_la_lai_doi and (
                re.search(r'^(lai|doi)', lujvo) or\
                re.search(r'[aeiouy](lai|doi)', lujvo) or\
                re.search(r'^la[^u]', lujvo) or\
                re.search(r'[aeiouy]la[^u]', lujvo)
            )
        )

    answersf = [x for x in answers if la_lai_doi(x['lujvo'])]
    return sorted(answersf, key=lambda an: an['score'])


def normalize(rafsi_list):
    '''
    rafsi_listにlujvoとして繋げる際に必要なhyphen文字を補完
    '''
    rafsis = rafsi_list[:]
    result = [rafsis.pop()]
    while rafsis:
        rafsi = rafsis.pop()
        rafsi_cv = get_cv(rafsi)

        end = rafsi[-1]
        init = result[0][0]

        if rafsi_cv == 'CVCC' or rafsi_cv == 'CCVC':
            result.insert(0, 'y')
        elif get_cv(end) == 'C' \
            and get_cv(init) == 'C' \
            and permissible_c(end + init) == 0:
            result.insert(0, 'y')
        elif end == 'n' and result[0][:2] in ['ts', 'tc', 'dz', 'dj']:
            result.insert(0, 'y')
        elif not rafsis:
            if rafsi_cv == 'CVV' or rafsi_cv == "CV'V":
                if len(rafsi_list) > 2 or not get_cv(result[0]) == 'CCV':
                    hyphen = 'n' if result[0][0] == 'r' else 'r'
                    result.insert(0, hyphen)

            elif rafsi_cv == 'CVC' and tosmabru(rafsi, result):
                result.insert(0, 'y')

        result.insert(0, rafsi)
    return result

# tosmabru test
# https://mw.lojban.org/papri/tosmabru
def tosmabru(rafsi, rest):
    '''
    tosmabru test
    '''
    if is_cmevla(rest[-1]):
        return

    index = 0
    for i, resti in enumerate(rest):
        resti_cv = get_cv(resti)
        if resti_cv == 'CVC':
            continue
        index = i
        if resti == 'y' or (resti_cv == 'CVCCV' and permissible_c(resti[2:4]) == 2):
            break
        else:
            return False

    tmp1 = rafsi
    tmp2 = rest[0]
    j = 0
    while True:
        if tmp2 == 'y':
            return True
        if permissible_c(tmp1[-1] + tmp2[0]) != 2:
            return False

        j += 1
        tmp1 = tmp2
        tmp2 = rest[j]
        if j > index:
            break
    return True

def create_every_possibility(rafsi_list):
    '''
    可能性のある組み合わせを全て返す
    '''
    ary = rafsi_list[:]

    if not len(ary):
        return [[]]

    rafsis = ary.pop()
    result = []

    for rafs in rafsis:
        result += [f + [rafs] for f in create_every_possibility(ary)]
    return result


JSON_PATH = './data/origin/jbovlaste.json'
with open(JSON_PATH, 'r', encoding='utf-8') as text:
    DICTIONARY = json.loads(text.read())

def get_candid(valsi, islast=False):
    '''
    get_candid("bloti", false) ==> ["lot", "blo", "lo'i", "blot"]
    get_candid("gismu", true)  ==> ["gim", "gi'u", "gismu", "gism"]
    '''
    for data in DICTIONARY['words']:
        if data['word'] == valsi:
            if data['type'] == 'cmavo':
                return list(data['rafsi'].values())
            elif data['type'] == 'gismu':
                gismu = valsi
                candid = list(data['rafsi'].values())

                if islast:
                    candid.append(gismu)

                gism = gismu[:-1]
                if gism != 'brod':
                    candid.append(gism)
                return candid
    print('not found')

if __name__ == '__main__':
    # if sys.argv[1]:
    print(create_every_possibility([[1, 11], [2], [3, 33, 333]]))
    print(get_candid("gismu", True))
    for answer in jvozba(['lujvo', 'zbasu']):
        print(answer['lujvo'], answer['score'])
